### PR TITLE
Segmant_max 

### LINKED ISSUE #25778: segment_max

### PR DESCRIPTION
**Pull request by abdul-rehman18**
# PR Description

Segment_max Function in Tensorflow frontend

## Related Issue

Close #25778 

## Checklist

- [x] Did you add a function?
- [x] Did you add the tests?
- [x] Did you follow the steps we provided?

### CODE REVIEW

**Comment by abdul-rehman18** [comment]
@jieunboy0516 are you reviewing my pr

**Comment by jieunboy0516** [comment]
@abdul-rehman18 Thanks for your contribution to ivy and sorry for the late reply. It had been a busy week. Your code looks good to me but the tests are failing due to `hypothesis.errors.InvalidArgument: Cannot sample from a length-zero sequence.` Could you double check that please? Thanks

**Comment by abdul-rehman18** [comment]
@jieunboy0516 Yes I'll double check it

**Comment by abdul-rehman18** [comment]
@jieunboy0516 has this pr been merged?

**Comment by jieunboy0516** [comment]
> @jieunboy0516 has this pr been merged?

sorry i am not allowed to merge it when there are issues with the unit test. But i will merge it asap after the unit tests are passing. Thanks a lot

**Comment by abdul-rehman18** [comment]
@jieunboy0516 ok i'll fix the issue right away

**Comment by abdul-rehman18** [comment]
@jieunboy0516 kindly check it now I made some modifications in test_math.py according to my understanding.

**Comment by jieunboy0516** [comment]
the unit tests are still having the same error. please follow the guide on [https://unify.ai/docs/ivy/overview/contributing/setting_up.html#setting-up-testing-in-pycharm](this doc) to confirm the tests are passing!

**Comment by abdul-rehman18** [comment]
@jieunboy0516 Sorry to bother you again but could you please assist me in one thing. I have tried testing and checks are ok. Could you please tell me which file is showing these errors. i added code in math.py and test_math.py and according to my understanding no other file needed modification. I doubled checked my function. Thanks

**Comment by jieunboy0516** [comment]
> @jieunboy0516 Sorry to bother you again but could you please assist me in one thing. I have tried testing and checks are ok. Could you please tell me which file is showing these errors. i added code in math.py and test_math.py and according to my understanding no other file needed modification. I doubled checked my function. Thanks

Hi, may i ask if you have run the unit tests and did the tests run successfully? Since i saw that the unit tests are not passed:
image

**Comment by abdul-rehman18** [comment]
@jieunboy0516 i am currently sorting this out will solve it out soon
Thanks

**Review by HaiderSultanArc — COMMENTED** [review]
TensorFlow, Jax, and PaddlePaddle have `segment_max` in their APIs. 
- TensorFlow: [tf.math.segment_max](https://www.tensorflow.org/api_docs/python/tf/math/segment_max)
- Jax: [jax.ops.segment_max](https://jax.readthedocs.io/en/latest/_autosummary/jax.ops.segment_max.html)
- PaddlePaddle: [paddle.incubate.segment_max](https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/incubate/segment_max_en.html)

You should try to implement these in the supported backends and functional API of Ivy, as [mixed functions](https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions)

Then implement this frontend function with `ivy.segment_max` 😄

**Comment by abdul-rehman18** [comment]
@jieunboy0516 I looked at the issue and one variable was causing the problem. I have corrected the function. kindly review it.
Thanks

**Comment by abdul-rehman18** [comment]
@jieunboy0516

File: ivy/functional/backends/jax/activations.py [hunk 1]
```diff
@@ -46,6 +46,24 @@ def sigmoid(
     return 1 / (1 + jnp.exp(-x))
 
 
+def segment_max(
```
Comment (jieunboy0516) [inline ivy/functional/backends/jax/activations.py:49]: Hi, is there any reason `jax.ops.segment_max` was not used?

File: ivy/functional/backends/jax/activations.py [hunk 2]
```diff
@@ -46,6 +46,24 @@ def sigmoid(
...
+    unique_segment_ids, segment_counts = jnp.unique(segment_ids, return_counts=True)
+    max_values = jax.vmap(lambda segment_id: jnp.max(data[segment_ids == segment_id]))(
+        unique_segment_ids
+    )
+
+    result = jax.ops.segment_sum(
+        jax.ops.index_add, jnp.zeros_like(data), max_values, segment_ids
+    )
+
+    if out is not None:
```
Comment (jieunboy0516) [inline ivy/functional/backends/jax/activations.py:61]: this block of code copying to `out` is not needed

File: ivy/functional/backends/tensorflow/activations.py [hunk 3]
```diff
@@ -46,6 +46,12 @@ def relu(x: Tensor, /, *, complex_mode="jax", out: Optional[Tensor] = None) -> T
     return tf.nn.relu(x)
 
 
+def segment_max(
+    x: Tensor, /, *, complex_mode="jax", out: Optional[Tensor] = None
```
Comment (jieunboy0516) [inline ivy/functional/backends/tensorflow/activations.py:50]: the argument `segment_ids` shouldnt be omitted

File: ivy/functional/frontends/tensorflow/math.py [hunk 4]
```diff
@@ -612,6 +612,28 @@ def scalar_mul(scalar, x, name="scalar_mul"):
     return ivy.multiply(x, scalar).astype(x.dtype)
 
 
+@to_ivy_arrays_and_back
+def segment_max(data, segment_ids, name="segment_max"):
+    data = ivy.array(data)
```
Comment (jieunboy0516) [inline ivy/functional/frontends/tensorflow/math.py:617]: in the frontend, you should just return `ivy.segment_max()`.
and you should add this block of code to `ivy.functional.ivy.activations.py` instead. please have a look at [https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions](https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions)

File: ivy/stateful/activations.py [hunk 5]
```diff
@@ -550,3 +550,25 @@ def _forward(self, x):
             The outputs following the LogSigmoid activation *[batch_shape, d]*
         """
         return ivy.logsigmoid(x, complex_mode=self._complex_mode)
+
+
+class SegmentMax(Module):
```
Comment (jieunboy0516) [inline ivy/stateful/activations.py:555]: stateful is not needed for maths functions

File: ivy/functional/backends/paddle/activations.py [hunk 6]
```diff
@@ -210,3 +210,27 @@ def hardswish(
...
+
+
+def segment_max(
+    data: paddle.Tensor,
+    segment_ids: paddle.Tensor,
+    /,
+    *,
+    out: Optional[paddle.Tensor] = None,
+) -> paddle.Tensor:
+    unique_segment_ids, segment_counts = paddle.unique(segment_ids)
```
Comment (jieunboy0516) [inline ivy/functional/backends/paddle/activations.py:222]: have you tried using `paddle.incubate.segment_max` instead?

**Comment by jieunboy0516** [comment]
Hi @abdul-rehman18 , thanks for your contribution. I have left some comments and please have a look. I know that a [mixed function](https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions) is quite annoying to be implemented, thanks for your time on it.

File: ivy/functional/backends/jax/activations.py (hunk 1 above)
Comment (abdul-rehman18) [reply ivy/functional/backends/jax/activations.py:49]: I thought that this could work but now I have added it

File: ivy/functional/backends/tensorflow/activations.py (hunk 3 above)
Comment (abdul-rehman18) [reply ivy/functional/backends/tensorflow/activations.py]: I have added `segment_ids`

File: ivy/functional/frontends/tensorflow/math.py (hunk 4 above)
Comment (abdul-rehman18) [reply ivy/functional/frontends/tensorflow/math.py]: I have made changes in `ivy.functional.ivy.activation.py` as you instructed

File: ivy/stateful/activations.py (hunk 5 above)
Comment (abdul-rehman18) [reply ivy/stateful/activations.py]: removed that function

File: ivy/functional/backends/paddle/activations.py (hunk 6 above)
Comment (abdul-rehman18) [reply ivy/functional/backends/paddle/activations.py]: Yes i have simple returned the function with correct arguments

**Comment by abdul-rehman18** [comment]
@jieunboy0516 

Made changes as you instructed
Thanks

**Comment by abdul-rehman18** [comment]
@jieunboy0516 
Any chances of merging?

**Comment by abdul-rehman18** [comment]
@jieunboy0516 
image
i am getting this message

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.
